File: ios/Ubapp/Games/Tag/TagProtocol.swift
import Foundation

enum TagVariant: String, CaseIterable {
    case classic, freeze, zombie, hotPotato, bomb
    var displayName: String {
        switch self {
        case .classic: "Classic"; case .freeze: "Freeze tag"
        case .zombie: "Zombie"; case .hotPotato: "Hot potato"; case .bomb: "Bomb"
        }
    }
    var tagline: String {
        switch self {
        case .classic: "Tagger transfers the role on contact."
        case .freeze: "Tagged players freeze. Teammates can unfreeze them."
        case .zombie: "Tagged players become it too. Last survivor wins."
        case .hotPotato: "It must tag before the timer runs out — or they lose."
        case .bomb: "Only it knows their role. Tag before the hidden timer ends."
        }
    }
    /// Default round length. Hot potato uses this as the per-tag countdown;
    /// bomb uses it as the hidden bomb timer.
    var duration: TimeInterval {
        switch self {
        case .classic, .freeze, .zombie: 5 * 60
        case .hotPotato: 30
        case .bomb: 3 * 60
        }
    }
    var hasEarlyEnd: Bool { self != .classic }
    var hidesIt: Bool { self == .bomb }
}

/// Line-oriented JSON messages dispatched on `type`. Mirrors lib/games/tag/tag_protocol.dart.
enum TagMessage {
    case hello(peerId: String, displayName: String)
    case start(variant: TagVariant, startingItId: String, startTimeMs: Int64,
               peerIds: [String], peerNames: [String: String],
               durationOverrideSec: Int?)
    case tag(taggerId: String, victimId: String, timeMs: Int64)
    case unfreeze(unfreezerId: String, victimId: String, timeMs: Int64)
    case end(reason: String, winnerId: String?)
    case tutorialCall
    case tutorialVote(voterId: String, yes: Bool)
    case tutorialState(isOpen: Bool, yesCount: Int, noCount: Int,
                       eligibleCount: Int, result: Bool?, tutorialShown: Bool)

    func encode() -> String {
        let obj = toJSON()
        return String(data: try! JSONSerialization.data(withJSONObject: obj), encoding: .utf8)!
    }

    private func toJSON() -> [String: Any] {
        switch self {
        case let .hello(peerId, displayName):
            return ["type": "hello", "peerId": peerId, "displayName": displayName]
        case let .start(variant, startingItId, startTimeMs, peerIds, peerNames, durationOverrideSec):
            var d: [String: Any] = ["type": "start", "variant": variant.rawValue,
                    "startingItId": startingItId, "startTimeMs": startTimeMs,
                    "peerIds": peerIds, "peerNames": peerNames]
            if let s = durationOverrideSec { d["durationOverrideSec"] = s }
            return d
        case let .tag(taggerId, victimId, timeMs):
            return ["type": "tag", "taggerId": taggerId, "victimId": victimId, "timeMs": timeMs]
        case let .unfreeze(unfreezerId, victimId, timeMs):
            return ["type": "unfreeze", "unfreezerId": unfreezerId, "victimId": victimId, "timeMs": timeMs]
        case let .end(reason, winnerId):
            return ["type": "end", "reason": reason, "winnerId": winnerId as Any]
        case .tutorialCall:
            return ["type": "tutorial_call"]
        case let .tutorialVote(voterId, yes):
            return ["type": "tutorial_vote", "voterId": voterId, "yes": yes]
        case let .tutorialState(isOpen, yesCount, noCount, eligibleCount, result, tutorialShown):
            return ["type": "tutorial_state", "isOpen": isOpen,
                    "yesCount": yesCount, "noCount": noCount, "eligibleCount": eligibleCount,
                    "result": result as Any, "tutorialShown": tutorialShown]
        }
    }

    static func decode(_ raw: String) throws -> TagMessage {
        guard let data = raw.data(using: .utf8),
              let j = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = j["type"] as? String
        else { throw NSError(domain: "TagMessage", code: 0) }
        switch type {
        case "hello":
            return .hello(peerId: j["peerId"] as! String, displayName: j["displayName"] as! String)
        case "start":
            return .start(
                variant: TagVariant(rawValue: j["variant"] as! String) ?? .classic,
                startingItId: j["startingItId"] as! String,
                startTimeMs: (j["startTimeMs"] as! NSNumber).int64Value,
                peerIds: j["peerIds"] as! [String],
                peerNames: (j["peerNames"] as? [String: String]) ?? [:],
                durationOverrideSec: (j["durationOverrideSec"] as? NSNumber)?.intValue)
        case "tag":
            return .tag(taggerId: j["taggerId"] as! String, victimId: j["victimId"] as! String,
                        timeMs: (j["timeMs"] as! NSNumber).int64Value)
        case "unfreeze":
            return .unfreeze(unfreezerId: j["unfreezerId"] as! String, victimId: j["victimId"] as! String,
                             timeMs: (j["timeMs"] as! NSNumber).int64Value)
        case "end":
            return .end(reason: j["reason"] as! String, winnerId: j["winnerId"] as? String)
        case "tutorial_call":
            return .tutorialCall
        case "tutorial_vote":
            return .tutorialVote(voterId: j["voterId"] as! String, yes: j["yes"] as! Bool)
        case "tutorial_state":
            return .tutorialState(
                isOpen: j["isOpen"] as! Bool,
                yesCount: j["yesCount"] as! Int, noCount: j["noCount"] as! Int,
                eligibleCount: j["eligibleCount"] as! Int,
                result: j["result"] as? Bool,
                tutorialShown: j["tutorialShown"] as! Bool)
        default: throw NSError(domain: "TagMessage", code: 1)
        }
    }
}

File: android/app/src/main/java/com/example/jamboree/games/tag/TagProtocol.kt
package com.example.ubapp.games.tag

import org.json.JSONArray
import org.json.JSONObject

enum class TagVariant(
    val displayName: String,
    val tagline: String,
    /** Default round length in ms. Hot potato uses this as the per-tag countdown;
     *  bomb uses it as the hidden bomb timer. */
    val durationMs: Long,
) {
    CLASSIC("Classic", "Tagger transfers the role on contact.", 5 * 60_000L),
    FREEZE("Freeze tag", "Tagged players freeze. Teammates can unfreeze them.", 5 * 60_000L),
    ZOMBIE("Zombie", "Tagged players become it too. Last survivor wins.", 5 * 60_000L),
    HOT_POTATO("Hot potato", "It must tag before the timer runs out — or they lose.", 30_000L),
    BOMB("Bomb", "Only it knows their role. Tag before the hidden timer ends.", 3 * 60_000L);

    val hasEarlyEnd: Boolean get() = this != CLASSIC
    val hidesIt: Boolean get() = this == BOMB
}

/** Line-oriented JSON messages dispatched on `type`. */
sealed class TagMessage {
    abstract fun toJson(): JSONObject
    fun encode(): String = toJson().toString()

    data class Hello(val peerId: String, val displayName: String) : TagMessage() {
        override fun toJson() = JSONObject().put("type", "hello").put("peerId", peerId).put("displayName", displayName)
    }
    data class Start(
        val variant: TagVariant, val startingItId: String, val startTimeMs: Long,
        val peerIds: List<String>, val peerNames: Map<String, String>,
        val durationOverrideSec: Int? = null,
    ) : TagMessage() {
        override fun toJson() = JSONObject().apply {
            put("type", "start"); put("variant", variant.name.lowercase())
            put("startingItId", startingItId); put("startTimeMs", startTimeMs)
            put("peerIds", JSONArray(peerIds))
            put("peerNames", JSONObject(peerNames as Map<*, *>))
            durationOverrideSec?.let { put("durationOverrideSec", it) }
        }
    }
    data class Tag(val taggerId: String, val victimId: String, val timeMs: Long) : TagMessage() {
        override fun toJson() = JSONObject()
            .put("type", "tag").put("taggerId", taggerId).put("victimId", victimId).put("timeMs", timeMs)
    }
    data class Unfreeze(val unfreezerId: String, val victimId: String, val timeMs: Long) : TagMessage() {
        override fun toJson() = JSONObject()
            .put("type", "unfreeze").put("unfreezerId", unfreezerId).put("victimId", victimId).put("timeMs", timeMs)
    }
    data class End(val reason: String, val winnerId: String?) : TagMessage() {
        override fun toJson() = JSONObject().put("type", "end").put("reason", reason)
            .put("winnerId", winnerId ?: JSONObject.NULL)
    }
    object TutorialCall : TagMessage() {
        override fun toJson() = JSONObject().put("type", "tutorial_call")
    }
    data class TutorialVote(val voterId: String, val yes: Boolean) : TagMessage() {
        override fun toJson() = JSONObject()
            .put("type", "tutorial_vote").put("voterId", voterId).put("yes", yes)
    }
    data class TutorialState(
        val isOpen: Boolean, val yesCount: Int, val noCount: Int,
        val eligibleCount: Int, val result: Boolean?, val tutorialShown: Boolean,
    ) : TagMessage() {
        override fun toJson() = JSONObject()
            .put("type", "tutorial_state").put("isOpen", isOpen)
            .put("yesCount", yesCount).put("noCount", noCount)
            .put("eligibleCount", eligibleCount)
            .put("result", result ?: JSONObject.NULL)
            .put("tutorialShown", tutorialShown)
    }

    companion object {
        fun decode(raw: String): TagMessage = decode(JSONObject(raw))

        /** Decode from an already-parsed object — used by the app-peer join
         *  path, which carries Tag messages over the same `GuestLink` JSON
         *  channel browser-tier games use. */
        fun decode(j: JSONObject): TagMessage {
            return when (j.getString("type")) {
                "hello" -> Hello(j.getString("peerId"), j.getString("displayName"))
                "start" -> {
                    val ids = j.getJSONArray("peerIds").let { arr -> List(arr.length()) { arr.getString(it) } }
                    val names = j.optJSONObject("peerNames")?.let { obj ->
                        obj.keys().asSequence().associateWith { obj.getString(it) }
                    } ?: emptyMap()
                    Start(
                        variant = TagVariant.entries.first { it.name.equals(j.getString("variant"), true)
                            || it.name.replace("_", "").equals(j.getString("variant"), true) },
                        startingItId = j.getString("startingItId"),
                        startTimeMs = j.getLong("startTimeMs"),
                        peerIds = ids, peerNames = names,
                        durationOverrideSec = if (j.has("durationOverrideSec") && !j.isNull("durationOverrideSec"))
                            j.getInt("durationOverrideSec") else null,
                    )
                }
                "tag" -> Tag(j.getString("taggerId"), j.getString("victimId"), j.getLong("timeMs"))
                "unfreeze" -> Unfreeze(j.getString("unfreezerId"), j.getString("victimId"), j.getLong("timeMs"))
                "end" -> End(j.getString("reason"), if (j.isNull("winnerId")) null else j.getString("winnerId"))
                "tutorial_call" -> TutorialCall
                "tutorial_vote" -> TutorialVote(j.getString("voterId"), j.getBoolean("yes"))
                "tutorial_state" -> TutorialState(
                    isOpen = j.getBoolean("isOpen"),
                    yesCount = j.getInt("yesCount"), noCount = j.getInt("noCount"),
                    eligibleCount = j.getInt("eligibleCount"),
                    result = if (j.isNull("result")) null else j.getBoolean("result"),
                    tutorialShown = j.getBoolean("tutorialShown"),
                )
                else -> throw IllegalArgumentException("Unknown TagMessage: ${j.getString("type")}")
            }
        }
    }
}
